src/layout/IHyprLayout.cpp

```diff
@@ -396,6 +396,139 @@ void IHyprLayout::onEndDragWindow() {
     g_pInputManager->m_bWasDraggingWindow = false;
 }
 
+inline bool canSnap(const double sideA, const double sideB, const double gap) {
+    return std::abs(sideA - sideB) < gap;
+}
+
+void snapMoveLeft(double& pos, double& len, const double p) {
+    pos = p;
+}
+void snapMoveRight(double& pos, double& len, const double p) {
+    pos = p - len;
+}
+
+void snapResizeLeft(double& pos, double& len, const double p) {
+    len += pos - p;
+    pos = p;
+}
+void snapResizeRight(double& pos, double& len, const double p) {
+    len = p - pos;
+}
+
+typedef void (*SnapFn)(double& pos, double& len, const double p);
+
+void IHyprLayout::performSnap(Vector2D& pos, Vector2D& size, PHLWINDOW DRAGGINGWINDOW, const eMouseBindMode mode) {
+    static auto SNAPWINDOWGAP  = CConfigValue<Hyprlang::INT>("general:snap:window_gap");
+    static auto SNAPMONITORGAP = CConfigValue<Hyprlang::INT>("general:snap:monitor_gap");
+
+    int         corner, snaps = 0;
+    SnapFn      snapUp, snapDown, snapLeft, snapRight;
+    if (mode == MBIND_MOVE) {
+        corner = -1; // any corner
+        snapUp = snapLeft = snapMoveLeft;
+        snapDown = snapRight = snapMoveRight;
+    } else {
+        corner = m_eGrabbedCorner;
+        snapUp = snapLeft = snapResizeLeft;
+        snapDown = snapRight = snapResizeRight;
+    }
+
+    if (*SNAPWINDOWGAP) {
+        const auto PID  = DRAGGINGWINDOW->getPID();
+        const auto WSID = DRAGGINGWINDOW->workspaceID();
+        const int  BORD = DRAGGINGWINDOW->getRealBorderSize();
+
+        for (auto& other : g_pCompositor->m_vWindows) {
+            if (other->workspaceID() != WSID || other->getPID() == PID || !other->m_bIsMapped || other->m_bFadingOut || other->m_bIsX11)
```

Comment:
why are we disabling snapping to X11? Only OR should be skipped. `CWindow::isX11OverrideRedirect()`